vendor/vendor-apple/ios/src/main/kotlin/com/malinskiy/marathon/apple/ios/device/AppleSimulatorProvider.kt

```
@@ -136,40 +140,27 @@ class AppleSimulatorProvider(
-                                        } else {
-                                            throw RuntimeException("Device provider update too long: maximum is 2^20 bytes")
-                                        }
+                        while(true) {
+                            try {
+                                aSocket(ActorSelectorManager(Dispatchers.IO))
+                                    .tcp()
+                                    .connect(InetSocketAddress(dynamicConfiguration.host, dynamicConfiguration.port)).use { socket ->
+                                        readSocket(byteBufferWrapper, socket, this)
                                     }
-                                    readChannel.readFully(buffer)
-                                    buffer.compatRewind()
-                                    val devicesWithEnvironmentVariablesReplaced =
-                                        environmentVariableSubstitutor.replace(buffer.decodeString())
-                                    val marathonfile = try {
-                                        objectMapper.readValue<Marathondevices>(devicesWithEnvironmentVariablesReplaced)
-                                    } catch (e: JsonMappingException) {
-                                        throw NoDevicesException("Invalid Marathondevices update format", e)
-                                    }
-                                    send(marathonfile)
-                                }
+                            } catch (e: ConnectException) {
+                                logger.warn { "Connection refused, retrying in 5 seconds..." }
+                                delay(5000) // Retry delay if socket connection is refused
+                            } catch (e: Exception) {
+                                logger.error(e) { "Error occurred: ${e.message}" }
+                                break // Exit the reading loop if a critical error occurs
```

Should rethrow the exception to support coroutine cancellation guarantees